### ISSUE TITLE
vispy always using gpu0 for off-screen rendering. (egl-backend)

### ISSUE THREAD

**Issue by rfdeng** [vispy/vispy#2211]
Does Vispy have a way to specify which GPU to use when using egl backend. My container fails to use vispy when it's not mounted on gpu0.

my issue is much alike this one:
https://github.com/mmatl/pyrender/issues/92

And there is a solution in pyrender:
https://github.com/mmatl/pyrender/pull/60

I have tried using osmesa backend for off-screen rendering, but it's too slow so i have to choose the egl.

**Comment by djhoese** [comment]
I'm not sure we have a way of doing this unless it is by accident through what EGL already offers in their library. Here is the relevant bits I could find:

https://github.com/vispy/vispy/blob/5d58b832c853c45b44f5036b9e7c90d6e9f4f454/vispy/app/backends/_egl.py#L33

Which has this EGL documentation:

https://www.khronos.org/registry/EGL/sdk/docs/man/html/eglGetDisplay.xhtml

And is defined in vispy's ctypes wrapper here:

https://github.com/vispy/vispy/blob/5d58b832c853c45b44f5036b9e7c90d6e9f4f454/vispy/ext/egl.py#L235-L240

I don't fully understand, but I would be open to an environment variable being added if needed to switch to a different device/display number.

**Comment by rfdeng** [comment]
thanks for reply. @djhoese 
I have checked both python scripts you mentioned above. 
And I write a test script like this:

import ctypes
from ctypes import c_int as _c_int, POINTER as _POINTER, c_void_p, c_char_p

lib = ctypes.CDLL('/usr/lib64/libEGL.so.1')  # my libEGL.so.1 path.
lib.eglGetDisplay.restype = c_void_p
lib.eglInitialize.argtypes = c_void_p, _POINTER(_c_int), _POINTER(_c_int)

display = lib.eglGetDisplay(0)
print(display)
majorVersion = (_c_int*1)()
minorVersion = (_c_int*1)()
res = lib.eglInitialize(display, majorVersion, minorVersion)
print(bool(res))

I got True on virtual container which is mounted on gpu0, and got False when the virtual container mounted on other gpus.
And I haven't found a way to specify the gpu that can be used by EGL.
I would be appreciated if you could find such an environment variable that works.

**Comment by djhoese** [comment]
@larsoner Are you the "EGL guy"? Any ideas?

**Comment by larsoner** [comment]
No not really, I mostly just tried to get it working as backend IIRC by trial and error rather than useful knowledge. Never used it in practice...

**Comment by ansj11** [comment]
@rfdeng Hi, i also find that osmesa backend is too slow, but my egl backend always get fllowing error. I am running on a remote terminal, do you know the reason?

<img width="1180" alt="截屏2021-09-17 下午12 28 21" src="https://user-images.githubusercontent.com/24786133/134874806-fe0bb143-75f4-4960-ad75-61b68a05c55d.png">

**Comment by rfdeng** [comment]
> @rfdeng Hi, i also find that osmesa backend is too slow, but my egl backend always get fllowing error. I am running on a remote terminal, do you know the reason?
> 
> <img alt="截屏2021-09-17 下午12 28 21" width="1180" src="https://user-images.githubusercontent.com/24786133/134874806-fe0bb143-75f4-4960-ad75-61b68a05c55d.png">

@ansj11 It seems like your EGL hasn't installed or configured properly.

**Comment by djhoese** [comment]
@ansj11 I agree with @rfdeng. It looks like something isn't installed or configured properly. Regardless, let's not dilute the original conversation here with another issue. Let's move your discussion to your #2221 issue you created. This issue will be specifically about default display choice in EGL.

**Comment by ansj11** [comment]
Hi, now i am using osmesa to render a 82 frames video with 960x720 size, but it takes more than 500s
![截屏2021-09-28 下午2 29 24](https://user-images.githubusercontent.com/24786133/135034785-51ae24f0-9113-4712-9752-2af40eb1b70f.png)
